File: DFS/WordSearch/WordSearch.py
class Solution:
    def exist(self, board: list[list[str]], word: str) -> bool:
        col = len(board)
        row = len(board[0])

        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]

        def dfs(i, j, u):
            if board[i][j] != word[u]:
                return False
            if u == len(word) - 1:
                return True

            board[i][j] = "."

            for direction in directions:
                x = i + direction[0]
                y = j + direction[1]
                if 0 <= x < col and 0 <= y < row:
                    if dfs(x, y, u + 1):
                        return True

            board[i][j] = word[u]
            return False

        for i in range(col):
            for j in range(row):
                if dfs(i, j, 0):
                    return True

        return False

File: DFS/WordSearch/WordSearch.cs
public class Solution {
    public bool Exist(char[][] board, string word) {
        int col = board.Length;
        int row = board[0].Length;

        int[,] directions = new int[4,2] { { 1, 0 }, { -1, 0 }, { 0, 1 }, { 0, -1 } };

        bool dfs(int i, int j, int u)
        {
            if (board[i][j] == word[u])
                return false;
            if (u == word.Length - 1)
                return true;

            board[i][j] = ".";

            for (int k = 0; k < directions.GetLength(0); k++)
            {
                int x = i + direction[k, 0];
                int y = j + direction[k, 1];

                if (x >= 0 && x < col && y >= 0 && y < row)
                    if (dfs(x, y, u + 1))
                        return true;
            }

            board[i][j] = word[u];
            return false;
        }

        for (int i = 0; i < col; i++)
            for (int j = 0; j < row; j++)
            {
                if (dfs(i, j, 0))
                    return true;
            }

        return false;
    }
}